### ISSUE TITLE
Version bump

### ISSUE THREAD

**Issue by fredoboulo** [zeromq/jeromq#416]
Hello,

I'm about to propose a PR for a major upgrade to 4.1.7 (well, 4.1.6 plus 2 commits) 

This is not the most recent stable release but will help to reduce the gap between Jeromq and libzmq.

In short: ZMTP 1/2/3, Z85, ZAP, PLAIN, CURVE, Stream socket, ...

Let's go into the hard:

There will be no breaking API change, except for poll ( I removed the method with no selector, but it may be reintroduced) and for the next item in this list.

Custom encoder/decoder was removed from libzmq. I would propose to remove this feature as well: the same behaviour could be achieved with proxies.

All tests are passing, except the one regarding these custom decoder/encoder.

Performances are ok. Not better, not worse. I did not perform precise measurements.

Closing contexts... Still not the best situation (as of now). I will propose some work on that matter, but later.

I also took this opportunity to organize code into packages. I'm tired to see this long list of classes.
Ideally, only public API classes should belong to zmq: Ctx, SocketBase, ZMQ,...

My proposal:

zmq.io: all code related to IO threads
zmq.io.coder: the decoders/encoders
zmq.io.mechanism
zmq.io.net: transmission protocols
zmq.io.net.tcp
zmq.io.net.ipc
...
zmq.pipe: all about pipes
zmq.poll : polling mechanism
zmq.socket: all sockets would be listed here
zmq.socket.pipeline: push/pull
zmq.socket.pubsub
zmq.socket.reqrep: rep, req, dealer, router
...
zmq.util: utility classes

Unfortunately (or not), all code was reformatted. I will propose in the PR the eclipse profile that I used. It's conform to maven checkstyle.

I reorganized the code so it is closer to libzmq. Il would be easier to compare both versions, as libzmq was my reference (I still kept most of the tweaks for Java though)

It may be a bit scary (at least it's a bit scary to me), but I believe it will be a good opportunity for all of us.

I mostly tested on my environment (ubuntu), so feedbacks on other platforms would be greatly appreciated.

What do you feel about it?

**Comment by daveyarwood** [comment]
This sounds great to me! I've had the same thought about the long list of
classes, and wished that things were a little better organized. Organizing
our code as close as possible to the way libzmq is organized seems like a
good idea, so we can compare them side by side.

I'll be happy to review and test on macOS and Ubuntu.

**Comment by fredoboulo** [comment]
Cool :)

I will keep the flat version as long as possible to help reviewing.

I made a branch here : https://github.com/fredoboulo/jeromq/tree/proposal/v4.1.7.

There is still some work on the poller and the decoder, so I will wait for the PR

**Comment by daveyarwood** [comment]
Sorry I'm late to the party -- I see this has already been merged.

FWIW, here are the results when I run the tests in my Ubuntu 16.04 VM:

```
...

org.zeromq.TestZProxy
testAllOptionsSyncNewHot(org.zeromq.TestZProxy)  Time elapsed: 0.465 sec  <<< FAILURE!
org.junit.ComparisonFailure: sync status is not good! expected:<[STOPP]ED> but was:<[EXIT]ED>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at org.zeromq.TestZProxy.testAllOptionsSync(TestZProxy.java:298)
        at org.zeromq.TestZProxy.testAllOptionsSyncNewHot(TestZProxy.java:141)

testStateSyncPause(org.zeromq.TestZProxy)  Time elapsed: 0.339 sec  <<< FAILURE!
org.junit.ComparisonFailure: Start sync status is not good! expected:<[STAR]TED> but was:<[EXI]TED>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at org.zeromq.TestZProxy.testStateSyncPause(TestZProxy.java:384)
        at org.zeromq.TestZProxy.testStateSyncPause(TestZProxy.java:171)

...

Results :

Failed tests:
  TestZProxy.testAllOptionsSyncNewHot:141->testAllOptionsSync:298 sync status is not good! expected:<[STOPP]ED> but was:<[EXIT]ED>
  TestZProxy.testStateSyncPause:171->testStateSyncPause:384 Start sync status is not good! expected:<[STAR]TED> but was:<[EXI]TED>

Tests run: 150, Failures: 2, Errors: 0, Skipped: 8
```

And when I run them on my host system (macOS Sierra):

```
Results :

Tests run: 150, Failures: 0, Errors: 0, Skipped: 8
```

I think I've seen the 2 failing tests above fail intermittently in the past. There must be some race condition either in the tests or (worse) in the library causing things to go out of sync.

I did look at your code and I really like this effort to clean up and improve the code base. Major kudos!

**Comment by fredoboulo** [comment]
Not too late! 
I tried to improve the tests in #420 and #421, 'cause you're right they are randomly failing.
My last understanding of TestZProxy failing is that the port is chosen, but its use is deferred to a later moment. So there is a time window where an external process could use it in between. I changed that to remove this possibility of occurrence.
For TestProxy, I did move the sleep when I rewrote the test and did not let the time for sockets to establish connection before actual testing.

I also fixed some bugs, tried to improve the allocations, ...
Work is definitely not finished, I expect some more time for stabilization (and I honestly expected more feedbacks before merge).

May I ask if someone can make a code review of the immediate option (definition and use), compared to libzmq? The logic is... well, I would love to have feedback from fresh eyes.

I will propose more tests directly from libzmq and the known issues in Jeromq, after these 2 PR have been accepted.
